Public/Src/Cache/DistributedCache.Host/Service/Internal/CacheServerFactory.cs

```diff
@@ -0,0 +1,238 @@
+using BuildXL.Cache.ContentStore.Service;
+using BuildXL.Cache.ContentStore.Stores;
+using BuildXL.Cache.ContentStore.Utils;
+using BuildXL.Cache.Host.Configuration;
+using BuildXL.Cache.MemoizationStore.Interfaces.Caches;
+using BuildXL.Cache.MemoizationStore.Service;
+using BuildXL.Cache.MemoizationStore.Sessions;
+using BuildXL.Cache.MemoizationStore.Stores;
+
+namespace BuildXL.Cache.Host.Service.Internal
+{
+    /// <summary>
+    /// Creates and configures cache server instances.
+    /// </summary>
+    internal class CacheServerFactory
+    {
+        private readonly IAbsFileSystem _fileSystem;
+        private readonly ILogger _logger;
+        private readonly DistributedCacheServiceArguments _arguments;
+
+        public CacheServerFactory(DistributedCacheServiceArguments arguments)
+        {
+            _arguments = arguments;
+            _logger = arguments.Logger;
+            _fileSystem = new PassThroughFileSystem(_logger);
+        }
+
+        public StartupShutdownBase Create()
+        {
+            var cacheConfig = _arguments.Configuration;
+            cacheConfig.LocalCasSettings = cacheConfig.LocalCasSettings.FilterUnsupportedNamedCaches(_arguments.HostInfo.Capabilities, _logger);
+
+            var distributedSettings = cacheConfig.DistributedContentSettings;
+            var isLocal = distributedSettings == null || !distributedSettings.IsDistributedContentEnabled;
+
+            var serviceConfiguration = CreateServiceConfiguration(_logger, _fileSystem, cacheConfig.LocalCasSettings, new AbsolutePath(_arguments.DataRootPath), isDistributed: !isLocal);
+            var localServerConfiguration = CreateLocalServerConfiguration(cacheConfig.LocalCasSettings.ServiceSettings, serviceConfiguration);
+
+            if (isLocal)
+            {
+                return CreateLocalServer(localServerConfiguration);
+            }
+            else
+            {
+                return CreateDistributedServer(localServerConfiguration);
+            }
+        }
+
+        private StartupShutdownBase CreateLocalServer(LocalServerConfiguration localServerConfiguration)
+        {
+            Func<AbsolutePath, IContentStore> contentStoreFactory = path => ContentStoreFactory.CreateContentStore(_fileSystem, path, evictionAnnouncer: null, distributedEvictionSettings: default, contentStoreSettings: default, trimBulkAsync: null);
+
+            if (_arguments.EnableMetadataStore)
+            {
+                Func<AbsolutePath, ICache> cacheFactory = path => {
+                    return new OneLevelCache(
+                        contentStoreFunc: () => contentStoreFactory(path),
+                        memoizationStoreFunc: () => {
+                            return new RocksDbMemoizationStore(_logger, SystemClock.Instance, new RocksDbMemoizationStoreConfiguration()
+                            {
+                                Database = new RocksDbContentLocationDatabaseConfiguration(path / "RocksDbMemoizationStore") {

```

Comment:
`ContentStoreFactory.CreateContentStore` creates a `FileSystemContentStore`, which locks the directory. Is it correct to do what I am doing?

Reply:
It is fine.

---
In reply to: [314882874](https://github.com/microsoft/BuildXL/pull/751#discussion_r314882874) [](ancestors = 314882874)